### PR TITLE
[OPS-1711]: user agents to be rate limited

### PR DESCRIPTION
**Pull request by bilalahmad99**
## Configuration Pull Request

Make sure that the following steps are done before merging
- [ ] @devops team member has commented with :+1:
- [ ] are you adding any new default values that need to be overriden when this goes live?  
  - [ ] Open a ticket (DEVOPS) to make sure that they have been added to secure vars.
  - [ ] Add an entry to the CHANGELOG.

### CODE REVIEW

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 1]
```diff
@@ -205,6 +208,7 @@ error_page {{ k }} {{ v }};
     {%- if EDXAPP_ENABLE_RATE_LIMITING -%}
     # Set Limit
     limit_req zone=cookies burst={{ EDXAPP_COURSES_REQUEST_BURST_RATE }};
+    limit_req zone=agents burst={{ EDXAPP_COURSES_USER_AGENT_BURST_RATE }};
```
Comment (fredsmith) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:219]: You need to set a default value for this new variable

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 2]
```diff
@@ -14,6 +14,9 @@ upstream lms-backend {
 {%- if EDXAPP_ENABLE_RATE_LIMITING -%}
 # Make Zone
 limit_req_zone $cookie_{{ EDXAPP_SESSION_COOKIE_NAME }} zone=cookies:10m rate={{ EDXAPP_COURSES_REQUEST_RATE }};
+{% for agent in EDXAPP_COURSES_USER_AGENTS %}
+limit_req_zone {{ agent.agent_name }} $binary_remote_addr zone=agents:10m rate={{ agent.rate }};
```
Comment (fredsmith) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2]: shouldn't this be "$http_user_agent" and not the agent name?

Comment (fredsmith) [reply playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2]: Ah, ok, after reading the documentation on limit_req_zone, it looks like you're correct.

**Comment by fredsmith** [comment]
👍  after adding a default for EDXAPP_COURSES_USER_AGENT_BURST_RATE

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 (hunk 1 above)
Comment (bilalahmad99) [reply playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:219]: done

**Comment by bilalahmad99** [comment]
@edx/devops plz review again. found a few issues with previous solution while testing. I have added a new commit that fixes it now.

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 3]
```diff
@@ -14,6 +14,14 @@ upstream lms-backend {
 {%- if EDXAPP_ENABLE_RATE_LIMITING -%}
 # Make Zone
 limit_req_zone $cookie_{{ EDXAPP_SESSION_COOKIE_NAME }} zone=cookies:10m rate={{ EDXAPP_COURSES_REQUEST_RATE }};
+{% for agent in EDXAPP_COURSES_USER_AGENTS %}
```
Comment (jibsheet) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2]: Can we name this EDXAPP_RATE_LIMITED_USER_AGENTS instead (so it's clearer what this list is for?)

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 4]
```diff
@@ -14,6 +14,14 @@ upstream lms-backend {
...
+map $http_user_agent {{ "$limit_bot_" ~ agent.alias }} {
```
Comment (fredsmith) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:21]: Can you put a comment here to describe what is happening?  Eg.  "we make a bucket named $limit_bot_agent_alias"
and on a line lower "add binary IP of the user to the bucket"

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 5]
```diff
@@ -14,6 +14,14 @@ upstream lms-backend {
...
+          {{ agent.agent_name }} $binary_remote_addr;
+      }
+limit_req_zone {{ "$limit_bot_" ~ agent.alias }} zone=agents:10m rate={{ agent.rate }};
+{% endfor %}
```
Comment (jibsheet) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:26]: Can we pull an expanded version of this off your test sandbox and add it as a comment in your PR.  It looks correct, but we want to make sure the output looks right.

**Comment by fredsmith** [comment]
PR Failed: https://travis-ci.org/edx/configuration/jobs/164906622#L4082

```
TASK: [nginx | Copying nginx configs for ['lms', 'cms', 'xqueue', 'certs', 'forum']] *** 
fatal: [127.0.0.1] => {'msg': "AnsibleError: file: /edx/app/edx_ansible/edx_ansible/docker/plays/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2, line number: 25, error: Encountered unknown tag 'endif'.", 'failed': True}
fatal: [127.0.0.1] => {'msg': 'One or more items failed.', 'failed': True, 'changed': False, 'results': [{'msg': "AnsibleError: file: /edx/app/edx_ansible/edx_ansible/docker/plays/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2, line number: 25, error: Encountered unknown tag 'endif'.", 'failed': True}]}
```

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 [hunk 6]
```diff
@@ -14,6 +14,14 @@ upstream lms-backend {
...
 # Make Zone
 limit_req_zone $cookie_{{ EDXAPP_SESSION_COOKIE_NAME }} zone=cookies:10m rate={{ EDXAPP_COURSES_REQUEST_RATE }};
+{% for agent in EDXAPP_RATE_LIMITED_USER_AGENTS %}
+map $http_user_agent {{ "$limit_bot_" ~ agent.alias }} {
+          {{ agent.agent_name }} $binary_remote_addr;
+      }
+limit_req_zone {{ "$limit_bot_" ~ agent.alias }} zone=agents:10m rate={{ agent.rate }};
+{% endfor %}
+
+{%- endif %}
```
Comment (fredsmith) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:28]: extra endif

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2 (hunk 5 above)
Comment (bilalahmad99) [reply playbooks/roles/nginx/templates/edx/app/nginx/sites-available/lms.j2:26]: @fredsmith @jibsheet 

```
# Make Zone
limit_req_zone $cookie_sessionid zone=cookies:10m rate=5r/s;


# Map of http user agent with name limit_bot_agent_alias having binary IP of the agent
map $http_user_agent $limit_bot_edx_dl {
          edX-downloader $binary_remote_addr;
      }

limit_req_zone $limit_bot_edx_dl zone=agents:10m rate=5r/m;
```

**Review by fredsmith — APPROVED** [review]

**Pull request merged**